src/som/interpreter/nodes/dispatch/BackCacheCallNode.java

```diff
@@ -0,0 +1,47 @@
+package som.interpreter.nodes.dispatch;
+
+import com.oracle.truffle.api.frame.VirtualFrame;
+import com.oracle.truffle.api.nodes.Node;
+
+import som.interpreter.Invokable;
+import som.interpreter.SArguments;
+import som.vm.VmSettings;
+import tools.asyncstacktraces.ShadowStackEntry;
+import tools.asyncstacktraces.ShadowStackEntryLoad;
+
+
+public interface BackCacheCallNode {
+
+  static void setShadowStackEntry(final VirtualFrame frame,
+      final boolean uniqueCaller, final Object[] arguments,
+      final Node expression,
+      final ShadowStackEntryLoad shadowStackEntryLoad) {
+    if (VmSettings.ACTOR_ASYNC_STACK_TRACE_STRUCTURE) {
+      assert arguments[arguments.length - 1] == null;
+      assert (frame.getArguments()[frame.getArguments().length
+          - 1] instanceof ShadowStackEntry);
+      assert frame.getArguments().length >= 2;
+    }
+    if (VmSettings.ACTOR_ASYNC_STACK_TRACE_METHOD_CACHE) {
+      if (uniqueCaller) {
+        SArguments.setShadowStackEntry(arguments, SArguments.getShadowStackEntry(frame));
+      } else {
+        SArguments.setShadowStackEntryWithCache(arguments, expression,
+            shadowStackEntryLoad, frame, false);
+      }
+    } else if (VmSettings.ACTOR_ASYNC_STACK_TRACE_STRUCTURE) {
+      SArguments.setShadowStackEntryWithCache(arguments, expression,
+          shadowStackEntryLoad, frame, false);
+    }
+    assert arguments[arguments.length - 1] != null
+        || (frame.getArguments()[frame.getArguments().length - 1] == null)
+        || !VmSettings.ACTOR_ASYNC_STACK_TRACE_STRUCTURE;
+  }
+
+  void uniqueCaller();
```

Comment:
This and the next method seem to be mutating methods. Perhaps to be clear `set` or `make` would be a good prefix for them?